lib/Models/TableDataSource.js

```diff
@@ -0,0 +1,1008 @@
+    // Put the description of this row onto the image of the region.
+    var rowObjects = tableStructure.toRowObjects();
+    var regionRowObjects = regionIndices.map(function(i) { return rowObjects[i]; });
+    ImageryProviderHooks.addPickFeaturesHook(regionImageryProvider, function(results) {
+        if (!defined(results) || results.length === 0) {
+            return;
+        }
+        for (var i = 0; i < results.length; ++i) {
+            var uniqueId = results[i].data.properties[regionDetail.regionProvider.uniqueIdProp];
+            results[i].description = describeRow(regionRowObjects[uniqueId], dataSource._tableStyle);
+        }
+        return results;
+    });
+
+    dataSource._regionImageryLayer = ImageryLayerCatalogItem.enableLayer(catalogItem, regionImageryProvider, catalogItem.opacity, layerIndex);
+    if (defined(catalogItem.terria.leaflet) && colorFunction) {
+        dataSource._regionImageryLayer.setFilter(function () {
+            new L.CanvasFilter(this, {
+                channelFilter: function (image) {
+                    return ImageryProviderHooks.recolorImage(image, colorFunction);
+                }
+            }).render();
+        });
+    }
+}
+
+/**
+ * Update the region imagery layer, eg. when the active variable changes, or the time changes.
+ * Following previous practice, when the coloring needs to change, the item is hidden, disabled, then re-enabled and re-shown.
+ * So, a new imagery layer is created (during 'enable') each time its coloring changes.
+ * TODO: Can we reuse the existing one? (I found the recoloring doesn't get used.)
+ *
+ * @param  {TableDataSource} dataSource The data source.
+ * @param  {Array} [regionIndices] Passed into setNewRegionImageryLayer. Saves recalculating it if available.
+ */
+function redisplayRegions(dataSource, regionIndices) {
+    dataSource.regionPromise.then(function(regionDetails) {
+        if (regionDetails) {
+            // We are using _hadImageryAtLayerIndex = true to mean it had an ImageryLayer, but its layer was undefined.
+            // _hadImageryAtLayerIndex = undefined means it did not have an ImageryLayer.
+            var layerIndex = dataSource._hadImageryAtLayerIndex === true ? undefined : dataSource._hadImageryAtLayerIndex;
+            if (defined(dataSource._regionImageryLayer)) {
+                layerIndex = dataSource._regionImageryLayer._layerIndex;  // Would prefer not to access an internal variable of imageryLayer.
+                ImageryLayerCatalogItem.hideLayer(dataSource._catalogItem, dataSource._regionImageryLayer);
+                ImageryLayerCatalogItem.disableLayer(dataSource._catalogItem, dataSource._regionImageryLayer);
+                dataSource._regionImageryLayer = undefined;
+            }
+            setNewRegionImageryLayer(dataSource, layerIndex, regionIndices);
+            ImageryLayerCatalogItem.showLayer(dataSource._catalogItem, dataSource._regionImageryLayer);
+        }
+    });
```

Comment:
Nice catch, thanks Steve. All fixed now! In fact I've obliterated regionPromise in favour of a simpler approach - load() only resolves once regionDetails is available.